File: lib/core/security/booking_write.dart
import '../models/enums.dart';

/// Fail-closed booking writes. Matches live Dahr LY:
///
/// * **Consumer** — INSERT a pending request only. No `.update()` at all
///   (cancel is not a product path).
/// * **Vendor / admin** — may UPDATE. Vendor accept-with-quote stays on
///   [acceptRpcName]; decline / complete are direct status patches.
abstract final class BookingStatusWrite {
  static const String acceptRpcName = 'accept_booking_request';

  static const Set<BookingStatus> vendorDirectStatuses = {
    BookingStatus.declined,
    BookingStatus.completed,
  };

  /// Live RLS: couples cannot UPDATE `booking_requests`.
  static const bool consumerMayUpdate = false;

  /// Cancel was never a couple write. Do not add a consumer `.update()`.
  static const bool consumerMayCancel = false;

  static bool canInsert({
    required UserRole actor,
    required BookingStatus status,
  }) {
    if (actor == UserRole.consumer) {
      return status == BookingStatus.pending;
    }
    if (actor == UserRole.admin) return true;
    return false;
  }

  /// Any booking row UPDATE. Couples: never.
  static bool canUpdate({required UserRole actor}) {
    switch (actor) {
      case UserRole.consumer:
        return false;
      case UserRole.vendor:
      case UserRole.admin:
        return true;
    }
  }

  static bool canUpdateStatus({
    required UserRole actor,
    required BookingStatus status,
  }) {
    if (!canUpdate(actor: actor)) return false;
    switch (actor) {
      case UserRole.consumer:
        return false;
      case UserRole.vendor:
        return vendorDirectStatuses.contains(status);
      case UserRole.admin:
        return true;
    }
  }

  /// Consumer insert body. Rejects accepted/completed (or any non-pending).
  static Map<String, dynamic> consumerInsert(Map<String, dynamic> json) {
    final status = BookingStatus.fromString(json['status'] as String?);
    if (!canInsert(actor: UserRole.consumer, status: status)) {
      throw StateError('write_rejected');
    }
    return json;
  }

  /// Any couple UPDATE (status, cancel, quote, message) is rejected.
  static Map<String, dynamic> consumerUpdate([
    Map<String, dynamic>? _,
  ]) {
    throw StateError('write_rejected');
  }

  /// Status-only patch. Couples never get a map; vendors cannot accept here.
  static Map<String, dynamic> statusPatch({
    required UserRole actor,
    required BookingStatus status,
  }) {
    if (actor == UserRole.vendor && status == BookingStatus.accepted) {
      throw StateError('quoted_amount_required');
    }
    if (!canUpdateStatus(actor: actor, status: status)) {
      throw StateError('write_rejected');
    }
    return {'status': status.name};
  }

  static Map<String, dynamic> vendorDirectPatch(BookingStatus status) =>
      statusPatch(actor: UserRole.vendor, status: status);

  static Map<String, dynamic> adminModeratePatch(BookingStatus status) =>
      statusPatch(actor: UserRole.admin, status: status);

  static Map<String, dynamic> consumerStatusPatch(BookingStatus status) =>
      consumerUpdate({'status': status.name});
}

File: admin/src/lib/booking-write.ts
/** Fail-closed booking writes. Mirrors Flutter `BookingStatusWrite`. Live Dahr LY:
 * couples cannot UPDATE bookings at all. Vendor and admin still can.
 * Vendor accept-with-quote stays on `accept_booking_request`.
 */

export const BOOKING_STATUSES = [
  "pending",
  "accepted",
  "declined",
  "completed",
] as const;

export type BookingStatus = (typeof BOOKING_STATUSES)[number];
export type BookingWriteActor = "consumer" | "vendor" | "admin";

export const ACCEPT_BOOKING_RPC = "accept_booking_request";
export const VENDOR_DIRECT_STATUSES = ["declined", "completed"] as const;

/** Live RLS: couples cannot UPDATE `booking_requests`. */
export const CONSUMER_MAY_UPDATE = false;

/** Cancel is not a couple write. Do not add a consumer `.update()`. */
export const CONSUMER_MAY_CANCEL = false;

export function canInsertBooking(
  actor: BookingWriteActor,
  status: BookingStatus,
): boolean {
  if (actor === "consumer") return status === "pending";
  if (actor === "admin") return true;
  return false;
}

/** Any booking row UPDATE. Couples: never. */
export function canUpdateBooking(actor: BookingWriteActor): boolean {
  return actor === "vendor" || actor === "admin";
}

export function canUpdateBookingStatus(
  actor: BookingWriteActor,
  status: BookingStatus,
): boolean {
  if (!canUpdateBooking(actor)) return false;
  if (actor === "vendor") {
    return status === "declined" || status === "completed";
  }
  if (actor === "admin") return BOOKING_STATUSES.includes(status);
  return false;
}

export function consumerInsertStatus(status: BookingStatus): BookingStatus {
  if (!canInsertBooking("consumer", status)) {
    throw new Error("write_rejected");
  }
  return status;
}

/** Any couple UPDATE (status, cancel, quote, message) is rejected. */
export function consumerBookingUpdate(): never {
  throw new Error("write_rejected");
}

/**
 * Status-only patch. Couples never get a map. Vendor accept is the RPC, not
 * a row update. Admin may moderate any known status.
 */
export function bookingStatusPatch(
  actor: BookingWriteActor,
  status: BookingStatus,
): { status: BookingStatus } {
  if (actor === "consumer") {
    return consumerBookingUpdate();
  }
  if (actor === "vendor" && status === "accepted") {
    throw new Error("quoted_amount_required");
  }
  if (!canUpdateBookingStatus(actor, status)) {
    throw new Error("write_rejected");
  }
  return { status };
}

export function vendorDirectBookingPatch(status: BookingStatus) {
  return bookingStatusPatch("vendor", status);
}

export function adminModerateBookingPatch(status: BookingStatus) {
  return bookingStatusPatch("admin", status);
}
